pennylane/template.py

```diff
@@ -0,0 +1,225 @@
+# Copyright 2018 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# pylint: disable=protected-access
+r"""
+QML Templates
+=============
+
+**Module name:** :mod:`pennylane.template`
+
+.. currentmodule:: pennylane.template
+
+This module provides a growing library of templates of common quantum
+machine learning circuit architectures that can be used to easily build
+more complex quantum machine learning models.
+
+For example, you can construct, evaluate, and train quantum circuit based
+on the circuit-centric quantum classifier architecture from
+:cite:`schuld2018circuit` on an arbitrary number of wires and with an
+arbitrary number of blocks in the following way:
+
+.. code-block:: python
+
+    import pennylane as qml
+    from pennylane import numpy as np
+
+    num_wires=4
+    num_blocks=3
+    dev = qml.device('default.qubit', wires=num_wires)
+
+    @qml.qnode(dev)
+    def circuit(weights, x=None):
+        qml.BasisState(x, wires=range(num_wires))
+        qml.template.CircuitCentric(weights, periodic=True, wires=range(num_wires))
+        return qml.expval.PauliZ(0)
+
+    weights=np.random.randn(num_blocks, num_wires, 3)
+    print(circuit(weights, x=np.array(np.random.randint(0, 1, num_wires))))
+
+
+The handy :func:`Interferometer` function can be used to construct arbitrary
+interferometers in terms of elementary :class:`~.Beamsplitter` Operations.
```

Comment:
What about phase rotations in the interferometer? Are these accounted for?

Reply:
I am not sure I understand. You mean the phase from the path between the BSs?

Reply:
Yes, our beamsplitters are not general enough to implement SU(N) interferometers. For that, we also need phase shift gates (as you'll see in the decompositions). Thus, we should mention in the docstring that the interferometers are built from BSgates *and* Rgates

Reply:
I am still not 100% sure how to fix this. Inspecting again the Clements paper, I see that what they call a BS there is slightly different from what a BS in is SF and PL.

If I understand the [SF docs](https://strawberryfields.readthedocs.io/en/latest/conventions/gates.html#beamsplitter) correctly (and assuming that Eq. (1) from Clements is the matrix acting on the vector of annihilation operators $\hat{a}_i$), then the BS in SF/PL does have, instead of the $e^{i\phi}$ in front of the $cos(\theta)$ in Eq. (1) in Clemets' paper, a $e^{-i\phi}$ in front of the $-\sin(\theta)$.

_Side remark: If this is correct, then why does Clements call $\cos(\theta)$ the "reflectivity"? Shouldn't this be the "trasnmitivity"?_

Assuming that the above is indeed correct, then
```
T_01(theta, phi) 
= sf.BSgate(theta, phi) | (q[0], q[1])
  sf.Rgate(phi) | q[0]
= qml.Beamsplitter(theta, phi, wires=[1, 2])
  qml.Rotation(phi, wires=[1])
```
, i.e., "Clemets' BS on modes 0 and 1 is the same as a SF/PL BS on modes 0 and 1 followed by a Rotation on mode 0".  Is that true?

If all of the above is correct, then I think I should:

1) Add `Rotation(phi[gate_num], wires=[i])` after line 224 in `template.py`
2) change the docstring accordingly. 

Can you please double check and correct me if I am wrong? Thanks!!!

Reply:
@cgogolin, yes correct - since Clements BS convention is slightly different to the one chosen in SF, we build up our triangular decomposition by converting each of Clement's beamsplitters into a beamsplitter on modes [0, 1], followed by a rotation on mode 0

Reply:
Yes, @cgogolin that is my understanding as well. Go ahead, as you suggested, and add the missing rotation gates to the first output mode of all beamsplitters

Reply:
Done in 036ac6f059e86af4831f1824f9813e2050780e33